### PR TITLE
Basic tray icon implementation

### LINKED ISSUE #28: Stay in tray

### PR DESCRIPTION
**Pull request by p-try**
### ☑️ Resolves

#28 Stay in tray

### 🖼️ Screenshots

🏚️ Before | 🏡 After
---|---
./. | grafik

### 🚧 Tasks

- [X] Add tray icon
- [X] Main window minimizes to tray
- [ ] Clicking on notifications opens main window

### CODE REVIEW

**Comment by nickvergessen** [comment]
Nice stuff 🤩

**Comment by nickvergessen** [comment]
Hi @p-try ! :) We have a Nextcloud Talk instance where we invite Nextcloud contributors to be able to communicate more quickly, and we also have channels for e.g. "Talk app team". If you like, I could give you a guest account via the email address from your copyright (or any other email)?

**Comment by p-try** [comment]
Thank you very much, I'd like that!

**Review by marcoambrosini — COMMENTED** [review]
I'm testing this on Mac OS 13.2.1 and no tray icon is visible

**Comment by p-try** [comment]
> I'm testing this on Mac OS 13.2.1 and no tray icon is visible

There seem to be some special requirements for MacOS that I have not considered. I will implement them and get back to you for a test.

**Comment by nickvergessen** [comment]
@ShGKme you also wanted to leave a comment here, right?

**Review by ShGKme — CHANGES_REQUESTED** [review]
Hello @p-try,

Thank you for you contribution!

It works. I think, we may make the tray implementation less "basic" and then merge it.

Also, we use [ESLint](https://eslint.org) for checking formatting and code-style. Please, run `npm run lint` and fix linting errors.

**Comment by p-try** [comment]
> Hello @p-try,
> 
> Thank you for you contribution!
> 
> It works. I think, we may make the tray implementation less "basic" and then merge it.
> 
> Also, we use [ESLint](https://eslint.org) for checking formatting and code-style. Please, run `npm run lint` and fix linting errors.

Agreed. Sorry I wasn't able to follow up to this so far but I should find time shortly to work on this.

File: src/app/app.tray.js [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+} = require('electron')
+const path = require('path')
+
+/**
+ * Setup tray icon
+ *
+ * @param clickCb
+ */
+function setupTrayIcon(clickCb) {
+	const tray = new Tray(path.join(__dirname, '..', '..', 'img', 'icons', 'icon.png'))
```
[the remark this answers is not in the document]
Comment (p-try) [reply src/app/app.tray.js]: Currently, the icon will only work in development as the icon files are not yet copied to the build directory. I suggest CopyWebpackPlugin for this but I think this should be addressed in a separate discussion as this is about asset files in general and there may be more in future.

**Review by ShGKme — COMMENTED** [review]
Hi @p-try 

Thanks for your updates!

Could you rebase onto `main` and fix conflicts? And then we can merge it.

You would probably need to move the `mainWindow` setup to the end of the talk window creation on `appData:receive` (after line 157 in the current `main`).

https://github.com/nextcloud/talk-desktop/blob/32aa45a6e2317add5f9781a56360496eff28ee0a/src/main.js#L148-L157

App initialization is a bit changed now to handle and re-fetch invalid application data.

**Comment by p-try** [comment]
Thank you so much for your great guidance, @ShGKme!

I have found that I had to add the minimize-to-tray setup both in the `appData:receive handler` and the `authentication:login` handler because the talk window is initialized in both. Otherwise the tray icon would not work right after logging in but only on a fresh app instance when the authentication was skipped. I am not sure if my proposed solution for this is the best way as it appears a bit fiddly to me.

https://github.com/p-try/talk-desktop/blob/f0d464b1d31f7f424cd5f05ac77528b2de7350fb/src/app/app.tray.js#L60-L67

**Comment by ShGKme** [comment]
Hi @p-try,

> I have found that I had to add the minimize-to-tray setup both in the appData:receive handler and the authentication:login handler because the talk window is initialized in both. 

Yes, it makes sense, cool!

> I am not sure if my proposed solution for this is the best way as it appears a bit fiddly to me.

https://github.com/p-try/talk-desktop/blob/f0d464b1d31f7f424cd5f05ac77528b2de7350fb/src/app/app.tray.js#L60-L67

It's fine. We need to improve window management anyway, but after some point when it will be clear, what we want to do with all the windows, initializations, setups and etc.

---

Could you make a `rebase onto main`, not merge? Otherwise, this PR has all the commits now from Marth and all the changes in Files changes.

We use rebase to keep PR up to date on `master`/`main` in all our repositories.

Feel free to DM me in Talk with your guest account if you need any help.

**Comment by p-try** [comment]
Looks like this was a GitHub hiccup as I had rebased on main but it had still shown all the previous commits from when I had merged two weeks ago. Editing the PR to change the base branch to something else and then changing it back to main seems to have fixed the commit list.

**Comment by ShGKme** [comment]
@p-try would you mind if I rebase your branch and force push?

**Comment by ShGKme** [comment]
Hi @p-try 

I have updated your fork:
1. Drop the last commit with merge `origin/tray_icon` to `tray_icon`
2. Rebase onto the current main
3. `push --force-with-lease`

If you don't have any new local changes, you need to `fetch` and `reset origin/tray_icon` (not `pull`). If you have any new local changes, stash or cherry-pick commit.

The last problem DCO and commit author mismatch in https://github.com/nextcloud/talk-desktop/pull/79/commits/74ad3b8493971f29aea5aafb645c430e69f67333.

You need to update this commit description with interactive rebase and then do `git push --force-with-lease` again.

**Comment by p-try** [comment]
Thank you so much for your detailed instructions!

**Review by ShGKme — APPROVED** [review]
Thanks for your contribution, I think, we can merge it now :)

**Pull request merged**